File: frontend/src/api/readerAnnotations.ts
import { apiClient } from './client';

export interface ReaderBookmark { id: string; page_number: number; created_at: string; }
export interface ReaderNote { id: string; page_number: number; selected_text?: string | null; note_text: string; position_data?: Record<string, unknown> | null; created_at: string; updated_at: string; }
export interface ReaderHighlight { id: string; page_number: number; selected_text: string; color: string; position_data?: Record<string, unknown> | null; created_at: string; }

const path = (groupId: string, bookId: string, resource: string) => `/groups/${groupId}/books/${bookId}/${resource}`;

export const getReaderBookmarks = async (groupId: string, bookId: string) => (await apiClient.get<ReaderBookmark[]>(path(groupId, bookId, 'bookmarks'))).data;
export const createReaderBookmark = async (groupId: string, bookId: string, page_number: number) => (await apiClient.post<ReaderBookmark>(path(groupId, bookId, 'bookmarks'), { page_number })).data;
export const deleteReaderBookmark = async (groupId: string, bookId: string, id: string) => { await apiClient.delete(`${path(groupId, bookId, 'bookmarks')}/${id}`); };

export const getReaderNotes = async (groupId: string, bookId: string) => (await apiClient.get<ReaderNote[]>(path(groupId, bookId, 'notes'))).data;
export const createReaderNote = async (groupId: string, bookId: string, input: Pick<ReaderNote, 'page_number' | 'note_text'> & Partial<Pick<ReaderNote, 'selected_text' | 'position_data'>>) => (await apiClient.post<ReaderNote>(path(groupId, bookId, 'notes'), input)).data;
export const updateReaderNote = async (groupId: string, bookId: string, id: string, input: Partial<Pick<ReaderNote, 'note_text' | 'selected_text' | 'position_data'>>) => (await apiClient.patch<ReaderNote>(`${path(groupId, bookId, 'notes')}/${id}`, input)).data;
export const deleteReaderNote = async (groupId: string, bookId: string, id: string) => { await apiClient.delete(`${path(groupId, bookId, 'notes')}/${id}`); };

export const getReaderHighlights = async (groupId: string, bookId: string) => (await apiClient.get<ReaderHighlight[]>(path(groupId, bookId, 'highlights'))).data;
export const createReaderHighlight = async (groupId: string, bookId: string, input: Pick<ReaderHighlight, 'page_number' | 'selected_text'> & Partial<Pick<ReaderHighlight, 'color' | 'position_data'>>) => (await apiClient.post<ReaderHighlight>(path(groupId, bookId, 'highlights'), input)).data;
export const updateReaderHighlight = async (groupId: string, bookId: string, id: string, input: Partial<Pick<ReaderHighlight, 'color' | 'position_data'>>) => (await apiClient.patch<ReaderHighlight>(`${path(groupId, bookId, 'highlights')}/${id}`, input)).data;
export const deleteReaderHighlight = async (groupId: string, bookId: string, id: string) => { await apiClient.delete(`${path(groupId, bookId, 'highlights')}/${id}`); };

File: backend/app/api/v1/routes/reader_annotations.py
from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.deps import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.reader_annotations import (
    ReaderBookmarkCreate, ReaderBookmarkRead, ReaderHighlightCreate, ReaderHighlightRead,
    ReaderHighlightUpdate, ReaderNoteCreate, ReaderNoteRead, ReaderNoteUpdate, ReaderAnnotationsRead,
)
from app.services import reader_annotation_service

router = APIRouter(tags=["Reader annotations"])


@router.get("/groups/{group_id}/books/{book_id}/annotations", response_model=ReaderAnnotationsRead)
async def get_reader_annotations(group_id: str, book_id: str, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    bookmarks, notes, highlights = await reader_annotation_service.list_all_annotations(db, group_id, book_id, current_user.id)
    return ReaderAnnotationsRead(bookmarks=bookmarks, notes=notes, highlights=highlights)


@router.get("/groups/{group_id}/books/{book_id}/bookmarks", response_model=list[ReaderBookmarkRead])
async def list_reader_bookmarks(group_id: str, book_id: str, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await reader_annotation_service.list_bookmarks(db, group_id, book_id, current_user.id)


@router.post("/groups/{group_id}/books/{book_id}/bookmarks", response_model=ReaderBookmarkRead, status_code=status.HTTP_201_CREATED)
async def create_reader_bookmark(group_id: str, book_id: str, payload: ReaderBookmarkCreate, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await reader_annotation_service.create_bookmark(db, group_id, book_id, current_user.id, payload.page_number)


@router.delete("/groups/{group_id}/books/{book_id}/bookmarks/{bookmark_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_reader_bookmark(group_id: str, book_id: str, bookmark_id: str, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    await reader_annotation_service.delete_bookmark(db, group_id, book_id, bookmark_id, current_user.id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/groups/{group_id}/books/{book_id}/notes", response_model=list[ReaderNoteRead])
async def list_reader_notes(group_id: str, book_id: str, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await reader_annotation_service.list_notes(db, group_id, book_id, current_user.id)


@router.post("/groups/{group_id}/books/{book_id}/notes", response_model=ReaderNoteRead, status_code=status.HTTP_201_CREATED)
async def create_reader_note(group_id: str, book_id: str, payload: ReaderNoteCreate, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await reader_annotation_service.create_note(db, group_id, book_id, current_user.id, payload.model_dump())


@router.patch("/groups/{group_id}/books/{book_id}/notes/{note_id}", response_model=ReaderNoteRead)
async def update_reader_note(group_id: str, book_id: str, note_id: str, payload: ReaderNoteUpdate, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await reader_annotation_service.update_note(db, group_id, book_id, note_id, current_user.id, payload.model_dump(exclude_unset=True))


@router.delete("/groups/{group_id}/books/{book_id}/notes/{note_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_reader_note(group_id: str, book_id: str, note_id: str, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    await reader_annotation_service.delete_note(db, group_id, book_id, note_id, current_user.id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/groups/{group_id}/books/{book_id}/highlights", response_model=list[ReaderHighlightRead])
async def list_reader_highlights(group_id: str, book_id: str, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await reader_annotation_service.list_highlights(db, group_id, book_id, current_user.id)


@router.post("/groups/{group_id}/books/{book_id}/highlights", response_model=ReaderHighlightRead, status_code=status.HTTP_201_CREATED)
async def create_reader_highlight(group_id: str, book_id: str, payload: ReaderHighlightCreate, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await reader_annotation_service.create_highlight(db, group_id, book_id, current_user.id, payload.model_dump())


@router.patch("/groups/{group_id}/books/{book_id}/highlights/{highlight_id}", response_model=ReaderHighlightRead)
async def update_reader_highlight(group_id: str, book_id: str, highlight_id: str, payload: ReaderHighlightUpdate, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await reader_annotation_service.update_highlight(db, group_id, book_id, highlight_id, current_user.id, payload.model_dump(exclude_unset=True))


@router.delete("/groups/{group_id}/books/{book_id}/highlights/{highlight_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_reader_highlight(group_id: str, book_id: str, highlight_id: str, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    await reader_annotation_service.delete_highlight(db, group_id, book_id, highlight_id, current_user.id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
